src/Plutarch/Extra/Ord.hs

```diff
@@ -0,0 +1,280 @@
+ Copyright: (C) Liqwid Labs 2022
+ License: Apache 2.0
+ Maintainer: Koz Ross <koz@mlabs.city>
+ Portability: GHC only
+ Stability: Experimental
+
+ Ordering-related helpers and functionality.
+-}
+module Plutarch.Extra.Ord (
+    -- * Types
+    POrdering (..),
+    PComparator,
+
+    -- * Functions
+
+    -- ** Creating comparators
+    pfromOrd,
+    pfromOrdBy,
+
+    -- ** Combining comparators
+    pproductComparator,
+    psumComparator,
+
+    -- ** Transforming comparators
+    pmapComparator,
+    preverseComparator,
+
+    -- ** Using comparators
+    pcompareBy,
+    pequateBy,
+) where
+
+import Data.Semigroup (Semigroup (stimes), stimesIdempotentMonoid)
+import Plutarch.Internal.PlutusType (PlutusType (pcon', pmatch'))
+import Plutarch.Lift (
+    PConstantDecl (
+        PConstantRepr,
+        PConstanted,
+        pconstantFromRepr,
+        pconstantToRepr
+    ),
+    PUnsafeLiftDecl (PLifted),
+ )
+
+{- | A representation of a comparison at the Plutarch level. Equivalent to
+ 'Ordering' in Haskell.
+
+ @since 3.4.0
+-}
+data POrdering (s :: S)
+    = -- | Indicates a less-than relationship.
+      --
+      -- @since 3.4.0
+      PLT
+    | -- | Indicates equality.
+      --
+      -- @since 3.4.0
+      PEQ
+    | -- | Indicates a greater-than relationship.
+      --
+      -- @since 3.4.0
+      PGT
+    deriving stock
+        ( -- | @since 3.4.0
+          Show
+        )
+
+-- | @since 3.4.0
+instance PUnsafeLiftDecl POrdering where
+    type PLifted POrdering = Ordering
+
+-- | @since 3.4.0
+instance PConstantDecl Ordering where
+    type PConstantRepr Ordering = Integer
+    type PConstanted Ordering = POrdering
+    pconstantToRepr = \case
+        LT -> 0
+        EQ -> 1
+        GT -> 2
+    pconstantFromRepr = \case
+        0 -> pure LT
+        1 -> pure EQ
+        2 -> pure GT
+        _ -> Nothing
+
+-- | @since 3.4.0
+instance PlutusType POrdering where
+    type PInner POrdering = PInteger
+    pcon' = \case
+        PLT -> 0
+        PEQ -> 1
+        PGT -> 2
+    pmatch' x f =
+        pif
+            (x #== 0)
+            (f PLT)
+            ( pif
+                (x #== 1)
+                (f PEQ)
+                (f PGT)
+            )
+
+-- | @since 3.4.0
+instance PEq POrdering where
+    x #== y = pmatch x $ \case
+        PLT -> pmatch y $ \case
+            PLT -> pcon PTrue
+            _ -> pcon PFalse
+        PEQ -> pmatch y $ \case
+            PEQ -> pcon PTrue
+            _ -> pcon PFalse
+        PGT -> pmatch y $ \case
+            PGT -> pcon PTrue
+            _ -> pcon PFalse
+
+-- | @since 3.4.0
+instance Semigroup (Term s POrdering) where
+    x <> y = pmatch x $ \case
+        PLT -> pcon PLT
+        PEQ -> y
+        PGT -> pcon PGT
+    stimes = stimesIdempotentMonoid
+
+-- | @since 3.4.0
+instance Monoid (Term s POrdering) where
+    mempty = pcon PEQ
+
+-- TODO: PShow, PPartialOrd, POrd
+
+-- | @since 3.4.0
+newtype PComparator (a :: S -> Type) (s :: S)
```

Comment:
Ideally we have to change `POrd`...

Reply:
Ideally, but that can wait until we upstream.